File: src/tools/PowerTestBench.m
%
% Copyright (c) 2021-present LAAS-CNRS
%
%   This program is free software: you can redistribute it and/or modify
%   it under the terms of the GNU General Public License as published by
%   the Free Software Foundation, either version 2 of the License, or
%   (at your option) any later version.
%
% SPDX-License-Identifier: GPL-2.0-or-later
%

classdef PowerTestBench < handle
    % PowerTestBench  Safety-oriented API over a ThingSet transport.
    %
    % The supplied client must provide read(path) and write(path, values)
    % methods, as ThingSetTools does.

    properties (SetAccess = private)
        Client
    end

    properties (Constant)
        ModeIdle = 0
        ModePowerOn = 1
        ModePowerOff = 2
    end

    methods
        function obj = PowerTestBench(client)
            if nargin ~= 1 || ~ismethod(client, "read") || ~ismethod(client, "write")
                error("PowerTestBench:invalidClient", ...
                    "client must provide read(path) and write(path, values)");
            end
            obj.Client = client;
        end

        function modeValue = setMode(obj, mode)
            % Set and verify Config/Mode.
            modeValue = PowerTestBench.normalizeMode(mode);
            obj.writeAndVerify("Config", struct("Mode", modeValue));
        end

        function frequencyHz = setFrequency(obj, frequencyHz)
            % Set and verify Config/Frequency_Hz.
            frequencyHz = PowerTestBench.validateInteger( ...
                "frequencyHz", frequencyHz, 1, Inf);
            obj.writeAndVerify( ...
                "Config", struct("Frequency_Hz", frequencyHz));
        end

        function readback = configureLeg(obj, leg, settings)
            % Configure a leg from a scalar settings struct.
            %
            % Supported fields: enable, capacitor, driver, buck, boost,
            % dutyCycle, referenceValue, trackingVar, phaseShift,
            % deadTimeRisingNs, and deadTimeFallingNs.
            arguments
                obj
                leg
                settings (1,1) struct
            end

            legNumber = PowerTestBench.normalizeLeg(leg);
            allowed = [ ...
                "enable", "capacitor", "driver", "buck", "boost", ...
                "dutyCycle", "referenceValue", "trackingVar", ...
                "phaseShift", "deadTimeRisingNs", "deadTimeFallingNs"];
            PowerTestBench.rejectUnknownFields(settings, allowed, "leg setting");

            values = struct();
            boolMappings = {
                "enable", "wEnable";
                "capacitor", "wCapa";
                "driver", "wDriver";
                "buck", "wBuck";
                "boost", "wBoost"
            };
            for i = 1:size(boolMappings, 1)
                source = boolMappings{i, 1};
                target = boolMappings{i, 2};
                if isfield(settings, source)
                    values.(target) = PowerTestBench.validateLogical( ...
                        source, settings.(source));
                end
            end

            if isfield(settings, "buck") && isfield(settings, "boost") && ...
                    settings.buck && settings.boost
                error("PowerTestBench:validation", ...
                    "buck and boost cannot both be true");
            end
            if isfield(settings, "buck") && settings.buck
                values.wBoost = false;
            end
            if isfield(settings, "boost") && settings.boost
                values.wBuck = false;
            end

            if isfield(settings, "dutyCycle")
                duty = PowerTestBench.validateFinite( ...
                    "dutyCycle", settings.dutyCycle);
                if duty < 0 || duty > 1
                    error("PowerTestBench:validation", ...
                        "dutyCycle must be between 0 and 1");
                end
                values.wDutyCycle = duty;
            end
            if isfield(settings, "referenceValue")
                values.wReferenceValue = PowerTestBench.validateFinite( ...
                    "referenceValue", settings.referenceValue);
            end
            if isfield(settings, "trackingVar")
                values.wTrackingVar = char( ...
                    PowerTestBench.normalizeChannel(settings.trackingVar));
            end
            if isfield(settings, "phaseShift")
                values.wPhaseShift = PowerTestBench.validateInteger( ...
                    "phaseShift", settings.phaseShift, -360, 360);
            end
            if isfield(settings, "deadTimeRisingNs")
                values.wDeadTimeRising_ns = PowerTestBench.validateInteger( ...
                    "deadTimeRisingNs", settings.deadTimeRisingNs, 0, 65535);
            end
            if isfield(settings, "deadTimeFallingNs")
                values.wDeadTimeFalling_ns = PowerTestBench.validateInteger( ...
                    "deadTimeFallingNs", settings.deadTimeFallingNs, 0, 65535);
            end

            if isempty(fieldnames(values))
                error("PowerTestBench:validation", ...
                    "at least one leg setting must be supplied");
            end
            path = "Config/Leg" + legNumber;
            readback = obj.writeAndVerify(path, values, 5e-4);
        end

        function value = readLeg(obj, leg)
            % Read all configuration fields for one leg.
            legNumber = PowerTestBench.normalizeLeg(leg);
            value = obj.readObject("Config/Leg" + legNumber, "leg");
        end

        function value = readMeasurements(obj)
            % Read all live measurements.
            value = obj.readObject("Measurements", "measurement");
        end

        function value = readCalibration(obj, channel)
            % Read one sensor channel's calibration object.
            normalized = PowerTestBench.normalizeChannel(channel);
            value = obj.readObject("Calibration/" + normalized, "calibration");
        end

        function readback = setCalibration(obj, channel, settings)
            % Set gain/offset and optionally persist them with store=true.
            arguments
                obj
                channel
                settings (1,1) struct
            end
            normalized = PowerTestBench.normalizeChannel(channel);
            allowed = ["gain", "offset", "store"];
            PowerTestBench.rejectUnknownFields( ...
                settings, allowed, "calibration setting");

            values = struct();
            if isfield(settings, "gain")
                values.wGain = PowerTestBench.validateFinite( ...
                    "gain", settings.gain);
            end
            if isfield(settings, "offset")
                values.wOffset = PowerTestBench.validateFinite( ...
                    "offset", settings.offset);
            end
            if isfield(settings, "store")
                store = PowerTestBench.validateLogical("store", settings.store);
                if store
                    values.wStore = true;
                end
            end
            if isempty(fieldnames(values))
                error("PowerTestBench:validation", ...
                    "gain, offset, or store=true must be supplied");
            end

            expected = values;
            if isfield(values, "wStore")
                expected.wStore = false;
            end
            readback = obj.writeAndVerify( ...
                "Calibration/" + normalized, values, 5e-7, expected);
        end

        function metadata = readMetadata(obj)
            % Read converter metadata using host-friendly field names.
            raw = obj.readObject("Converter", "metadata");
            metadata = PowerTestBench.metadataFromReadback(raw);
        end

        function metadata = setMetadata(obj, settings)
            % Set and verify one or more persistent converter identity fields.
            arguments
                obj
                settings (1,1) struct
            end
            allowed = [ ...
                "boardName", "boardVersion", ...
                "serialNumber", "firmwareVersion"];
            PowerTestBench.rejectUnknownFields( ...
                settings, allowed, "metadata field");

            mappings = {
                "boardName", "wBoardName", 23;
                "boardVersion", "wBoardVersion", 15;
                "serialNumber", "wSerialNumber", 47;
                "firmwareVersion", "wFirmwareVersion", 31
            };
            values = struct();
            for i = 1:size(mappings, 1)
                source = mappings{i, 1};
                target = mappings{i, 2};
                maximumLength = mappings{i, 3};
                if isfield(settings, source)
                    values.(target) = PowerTestBench.validateMetadata( ...
                        source, settings.(source), maximumLength);
                end
            end
            if isempty(fieldnames(values))
                error("PowerTestBench:validation", ...
                    "at least one metadata field must be supplied");
            end

            raw = obj.writeAndVerify("Converter", values);
            metadata = PowerTestBench.metadataFromReadback(raw);
        end

        function powerOn(obj, leg, settings, options)
            % Safely configure and energize exactly one selected leg.
            arguments
                obj
                leg
                settings (1,1) struct = struct()
                options.ConnectDriver (1,1) logical = false
                options.ConnectCapacitor (1,1) logical = false
            end
            legNumber = PowerTestBench.normalizeLeg(leg);
            forbidden = intersect( ...
                fieldnames(settings), {'enable', 'driver', 'capacitor'});
            if ~isempty(forbidden)
                error("PowerTestBench:validation", ...
                    "powerOn controls these settings directly: %s", ...
                    strjoin(forbidden, ", "));
            end

            try
                obj.setMode(PowerTestBench.ModePowerOff);
                obj.configureLeg(1, struct("enable", false));
                obj.configureLeg(2, struct("enable", false));
                settings.enable = true;
                settings.driver = options.ConnectDriver;
                settings.capacitor = options.ConnectCapacitor;
                obj.configureLeg(legNumber, settings);
                obj.setMode(PowerTestBench.ModePowerOn);
            catch ME
                try
                    obj.shutdown();
                catch
                    % Keep the original power-on error.
                end
                rethrow(ME);
            end
        end

        function shutdown(obj, options)
            % Request POWER_OFF and disable both legs, attempting every step.
            arguments
                obj
                options.DisconnectHardware (1,1) logical = false
            end
            failures = strings(0, 1);

            try
                obj.setMode(PowerTestBench.ModePowerOff);
            catch ME
                failures(end+1) = string(ME.message);
            end

            for leg = 1:2
                settings = struct("enable", false);
                if options.DisconnectHardware
                    settings.driver = false;
                    settings.capacitor = false;
                end
                try
                    obj.configureLeg(leg, settings);
                catch ME
                    failures(end+1) = string(ME.message); %#ok<AGROW>
                end
            end

            if ~isempty(failures)
                error("PowerTestBench:shutdown", ...
                    "shutdown completed with errors: %s", ...
                    strjoin(failures, "; "));
            end
        end
    end

    methods (Access = private)
        function value = readObject(obj, path, label)
            try
                value = obj.Client.read(path);
            catch ME
                error("PowerTestBench:communication", ...
                    "ThingSet read failed at %s: %s", path, ME.message);
            end
            if ~isstruct(value) || ~isscalar(value)
                error("PowerTestBench:readback", ...
                    "%s readback is not a ThingSet object", label);
            end
        end

        function readback = writeAndVerify( ...
                obj, path, values, absoluteTolerance, expected)
            if nargin < 4
                absoluteTolerance = 0;
            end
            if nargin < 5
                expected = values;
            end

            try
                obj.Client.write(path, values);
                readback = obj.Client.read(path);
            catch ME
                error("PowerTestBench:communication", ...
                    "ThingSet operation failed at %s: %s", path, ME.message);
            end
            if ~isstruct(readback) || ~isscalar(readback)
                error("PowerTestBench:readback", ...
                    "expected object readback from %s", path);
            end

            names = fieldnames(expected);
            for i = 1:numel(names)
                name = names{i};
                if ~isfield(readback, name)
                    error("PowerTestBench:readback", ...
                        "%s/%s is missing from the firmware readback", ...
                        path, name);
                end
                if ~PowerTestBench.valuesMatch( ...
                        readback.(name), expected.(name), absoluteTolerance)
                    error("PowerTestBench:readback", ...
                        "%s/%s readback does not match the requested value", ...
                        path, name);
                end
            end
        end
    end

    methods (Static, Access = private)
        function legNumber = normalizeLeg(leg)
            if isnumeric(leg) && isscalar(leg) && isreal(leg) && ...
                    isfinite(leg) && fix(double(leg)) == double(leg) && ...
                    any(double(leg) == [1, 2])
                legNumber = double(leg);
                return
            end
            if ischar(leg) || (isstring(leg) && isscalar(leg))
                normalized = upper(string(leg));
                if normalized == "1" || normalized == "LEG1"
                    legNumber = 1;
                    return
                elseif normalized == "2" || normalized == "LEG2"
                    legNumber = 2;
                    return
                end
            end
            error("PowerTestBench:validation", ...
                "leg must be 1, 2, 'Leg1', or 'Leg2'");
        end

        function channel = normalizeChannel(channel)
            if ~(ischar(channel) || (isstring(channel) && isscalar(channel)))
                error("PowerTestBench:validation", ...
                    "channel must be V1, V2, VH, I1, I2, or IH");
            end
            channel = upper(string(channel));
            if ~any(channel == ["V1", "V2", "VH", "I1", "I2", "IH"])
                error("PowerTestBench:validation", ...
                    "channel must be V1, V2, VH, I1, I2, or IH");
            end
        end

        function modeValue = normalizeMode(mode)
            if ischar(mode) || (isstring(mode) && isscalar(mode))
                switch upper(string(mode))
                    case "IDLE"
                        modeValue = PowerTestBench.ModeIdle;
                    case "POWER_ON"
                        modeValue = PowerTestBench.ModePowerOn;
                    case "POWER_OFF"
                        modeValue = PowerTestBench.ModePowerOff;
                    otherwise
                        error("PowerTestBench:validation", ...
                            "mode must be IDLE, POWER_ON, POWER_OFF, 0, 1, or 2");
                end
                return
            end
            if isnumeric(mode) && isscalar(mode) && isreal(mode) && ...
                    isfinite(mode) && any(double(mode) == [0, 1, 2])
                modeValue = double(mode);
                return
            end
            error("PowerTestBench:validation", ...
                "mode must be IDLE, POWER_ON, POWER_OFF, 0, 1, or 2");
        end

        function value = validateLogical(name, value)
            if ~islogical(value) || ~isscalar(value)
                error("PowerTestBench:validation", "%s must be logical", name);
            end
        end

        function value = validateFinite(name, value)
            if ~isnumeric(value) || ~isscalar(value) || ~isreal(value) || ...
                    ~isfinite(value)
                error("PowerTestBench:validation", ...
                    "%s must be a finite number", name);
            end
            value = double(value);
        end

        function value = validateInteger(name, value, minimum, maximum)
            value = PowerTestBench.validateFinite(name, value);
            if fix(value) ~= value || value < minimum || value > maximum
                error("PowerTestBench:validation", ...
                    "%s must be an integer between %g and %g", ...
                    name, minimum, maximum);
            end
        end

        function value = validateMetadata(name, value, maximumLength)
            if ~(ischar(value) || (isstring(value) && isscalar(value)))
                error("PowerTestBench:validation", "%s must be text", name);
            end
            value = char(string(value));
            codes = double(value);
            if isempty(value) || numel(value) > maximumLength || ...
                    any(codes < 32 | codes > 126)
                error("PowerTestBench:validation", ...
                    "%s must contain 1 to %d printable ASCII characters", ...
                    name, maximumLength);
            end
        end

        function rejectUnknownFields(settings, allowed, description)
            unknown = setdiff(string(fieldnames(settings)), string(allowed));
            if ~isempty(unknown)
                error("PowerTestBench:validation", ...
                    "unknown %s: %s", description, strjoin(unknown, ", "));
            end
        end

        function matches = valuesMatch(actual, expected, absoluteTolerance)
            if isnumeric(expected) && ~islogical(expected)
                matches = isnumeric(actual) && isscalar(actual) && ...
                    isreal(actual) && isfinite(actual) && ...
                    abs(double(actual) - double(expected)) <= ...
                    max(absoluteTolerance, abs(double(expected)) * 1e-6);
            elseif ischar(expected) || isstring(expected)
                matches = (ischar(actual) || isstring(actual)) && ...
                    isscalar(string(actual)) && ...
                    string(actual) == string(expected);
            else
                matches = isequal(actual, expected);
            end
        end

        function metadata = metadataFromReadback(raw)
            required = [ ...
                "wBoardName", "wBoardVersion", ...
                "wSerialNumber", "wFirmwareVersion"];
            missing = required(~isfield(raw, cellstr(required)));
            if ~isempty(missing)
                error("PowerTestBench:readback", ...
                    "Converter metadata is missing: %s", ...
                    strjoin(missing, ", "));
            end
            metadata = struct( ...
                "boardName", string(raw.wBoardName), ...
                "boardVersion", string(raw.wBoardVersion), ...
                "serialNumber", string(raw.wSerialNumber), ...
                "firmwareVersion", string(raw.wFirmwareVersion));
        end
    end
end

File: src/tools/power_test_bench.py
#
# Copyright (c) 2021-present LAAS-CNRS
#
#   This program is free software: you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation, either version 2 of the License, or
#   (at your option) any later version.
#
# SPDX-License-Identifier: GPL-2.0-or-later
#

"""High-level, safety-oriented client for the ThingSet power test bench."""

from enum import IntEnum
import math


class TesterMode(IntEnum):
    """Values exposed by ``Config/Mode`` in the test-bench firmware."""

    IDLE = 0
    POWER_ON = 1
    POWER_OFF = 2


class PowerTestBenchError(RuntimeError):
    """Raised when validation, communication, or readback verification fails."""


class PowerTestBench:
    """Power-test-bench operations layered over a ThingSet transport.

    ``client`` may be :class:`thingset_tools.ThingSetTools` or any object
    providing compatible ``read(path)`` and ``write(path, values)`` methods.
    """

    _LEGS = {"1": 1, "LEG1": 1, "2": 2, "LEG2": 2}
    _CHANNELS = ("V1", "V2", "VH", "I1", "I2", "IH")
    _METADATA_FIELDS = {
        "board_name": ("wBoardName", 23),
        "board_version": ("wBoardVersion", 15),
        "serial_number": ("wSerialNumber", 47),
        "firmware_version": ("wFirmwareVersion", 31),
    }

    def __init__(self, client):
        if not callable(getattr(client, "read", None)) or not callable(
            getattr(client, "write", None)
        ):
            raise TypeError("client must provide read(path) and write(path, values)")
        self.client = client

    @classmethod
    def _normalize_leg(cls, leg):
        if isinstance(leg, bool):
            raise ValueError("leg must be 1, 2, 'Leg1', or 'Leg2'")
        key = str(leg).upper()
        try:
            return cls._LEGS[key]
        except KeyError as exc:
            raise ValueError("leg must be 1, 2, 'Leg1', or 'Leg2'") from exc

    @classmethod
    def _normalize_channel(cls, channel):
        normalized = str(channel).upper()
        if normalized not in cls._CHANNELS:
            raise ValueError(
                "channel must be one of " + ", ".join(cls._CHANNELS)
            )
        return normalized

    @staticmethod
    def _validate_bool(name, value):
        if not isinstance(value, bool):
            raise ValueError(f"{name} must be a bool")
        return value

    @staticmethod
    def _validate_finite(name, value):
        if isinstance(value, bool) or not isinstance(value, (int, float)):
            raise ValueError(f"{name} must be a finite number")
        value = float(value)
        if not math.isfinite(value):
            raise ValueError(f"{name} must be a finite number")
        return value

    @staticmethod
    def _validate_integer(name, value, minimum, maximum=None):
        if isinstance(value, bool) or not isinstance(value, int):
            raise ValueError(f"{name} must be an integer")
        if value < minimum or (maximum is not None and value > maximum):
            if maximum is None:
                raise ValueError(f"{name} must be at least {minimum}")
            raise ValueError(f"{name} must be between {minimum} and {maximum}")
        return value

    @staticmethod
    def _values_match(actual, expected, absolute_tolerance):
        if isinstance(expected, float):
            return (
                isinstance(actual, (int, float))
                and not isinstance(actual, bool)
                and math.isclose(
                    float(actual),
                    expected,
                    rel_tol=1e-6,
                    abs_tol=absolute_tolerance,
                )
            )
        return actual == expected

    def _write_and_verify(
        self, path, values, *, absolute_tolerance=0.0, expected=None
    ):
        try:
            self.client.write(path, values)
            readback = self.client.read(path)
        except Exception as exc:
            raise PowerTestBenchError(f"ThingSet operation failed at {path}: {exc}") from exc

        if not isinstance(readback, dict):
            raise PowerTestBenchError(
                f"expected object readback from {path}, got {readback!r}"
            )

        expected_values = values if expected is None else expected
        for name, requested in expected_values.items():
            if name not in readback:
                raise PowerTestBenchError(
                    f"{path}/{name} is missing from the firmware readback"
                )
            if not self._values_match(
                readback[name], requested, absolute_tolerance
            ):
                raise PowerTestBenchError(
                    f"{path}/{name} read back as {readback[name]!r}, "
                    f"requested {requested!r}"
                )
        return readback

    @staticmethod
    def _normalize_mode(mode):
        if isinstance(mode, TesterMode):
            return mode
        if isinstance(mode, str):
            try:
                return TesterMode[mode.upper()]
            except KeyError as exc:
                raise ValueError(
                    "mode must be IDLE, POWER_ON, POWER_OFF, 0, 1, or 2"
                ) from exc
        if isinstance(mode, bool):
            raise ValueError("mode must be IDLE, POWER_ON, POWER_OFF, 0, 1, or 2")
        try:
            return TesterMode(mode)
        except (TypeError, ValueError) as exc:
            raise ValueError(
                "mode must be IDLE, POWER_ON, POWER_OFF, 0, 1, or 2"
            ) from exc

    def set_mode(self, mode):
        """Set and verify the global tester mode."""

        normalized = self._normalize_mode(mode)
        self._write_and_verify("Config", {"Mode": int(normalized)})
        return normalized

    def set_frequency(self, frequency_hz):
        """Set and verify the PWM switching frequency."""

        frequency_hz = self._validate_integer(
            "frequency_hz", frequency_hz, minimum=1
        )
        self._write_and_verify("Config", {"Frequency_Hz": frequency_hz})
        return frequency_hz

    def configure_leg(
        self,
        leg,
        *,
        enable=None,
        capacitor=None,
        driver=None,
        buck=None,
        boost=None,
        duty_cycle=None,
        reference_value=None,
        tracking_var=None,
        phase_shift=None,
        dead_time_rising_ns=None,
        dead_time_falling_ns=None,
    ):
        """Validate, write, and verify one leg's supplied settings."""

        leg_number = self._normalize_leg(leg)
        values = {}

        boolean_fields = (
            ("wEnable", "enable", enable),
            ("wCapa", "capacitor", capacitor),
            ("wDriver", "driver", driver),
            ("wBuck", "buck", buck),
            ("wBoost", "boost", boost),
        )
        for field, argument, value in boolean_fields:
            if value is not None:
                values[field] = self._validate_bool(argument, value)

        if buck is True and boost is True:
            raise ValueError("buck and boost cannot both be true")
        if buck is True:
            values["wBoost"] = False
        if boost is True:
            values["wBuck"] = False

        if duty_cycle is not None:
            duty = self._validate_finite("duty_cycle", duty_cycle)
            if duty < 0.0 or duty > 1.0:
                raise ValueError("duty_cycle must be between 0 and 1")
            values["wDutyCycle"] = duty

        if reference_value is not None:
            values["wReferenceValue"] = self._validate_finite(
                "reference_value", reference_value
            )

        if tracking_var is not None:
            values["wTrackingVar"] = self._normalize_channel(tracking_var)

        if phase_shift is not None:
            values["wPhaseShift"] = self._validate_integer(
                "phase_shift", phase_shift, -360, 360
            )

        if dead_time_rising_ns is not None:
            values["wDeadTimeRising_ns"] = self._validate_integer(
                "dead_time_rising_ns", dead_time_rising_ns, 0, 65535
            )
        if dead_time_falling_ns is not None:
            values["wDeadTimeFalling_ns"] = self._validate_integer(
                "dead_time_falling_ns", dead_time_falling_ns, 0, 65535
            )

        if not values:
            raise ValueError("at least one leg setting must be supplied")

        return self._write_and_verify(
            f"Config/Leg{leg_number}", values, absolute_tolerance=5e-4
        )

    def read_leg(self, leg):
        """Read the complete configuration object for one leg."""

        leg_number = self._normalize_leg(leg)
        value = self.client.read(f"Config/Leg{leg_number}")
        if not isinstance(value, dict):
            raise PowerTestBenchError("leg readback is not a ThingSet object")
        return value

    def read_measurements(self):
        """Read all live measurements."""

        value = self.client.read("Measurements")
        if not isinstance(value, dict):
            raise PowerTestBenchError("measurement readback is not a ThingSet object")
        return value

    def read_calibration(self, channel):
        """Read gain, offset, and store state for one sensor channel."""

        normalized = self._normalize_channel(channel)
        value = self.client.read(f"Calibration/{normalized}")
        if not isinstance(value, dict):
            raise PowerTestBenchError("calibration readback is not a ThingSet object")
        return value

    def set_calibration(self, channel, *, gain=None, offset=None, store=False):
        """Set and verify one channel's calibration values.

        When ``store`` is true, firmware persists the values and resets
        ``wStore`` to false. The reset is verified; the firmware currently
        exposes no separate NVS status value.
        """

        normalized = self._normalize_channel(channel)
        values = {}
        if gain is not None:
            values["wGain"] = self._validate_finite("gain", gain)
        if offset is not None:
            values["wOffset"] = self._validate_finite("offset", offset)
        if not isinstance(store, bool):
            raise ValueError("store must be a bool")
        if store:
            values["wStore"] = True
        if not values:
            raise ValueError("gain, offset, or store=True must be supplied")

        expected = dict(values)
        if store:
            expected["wStore"] = False
        return self._write_and_verify(
            f"Calibration/{normalized}",
            values,
            absolute_tolerance=5e-7,
            expected=expected,
        )

    def read_metadata(self):
        """Return converter metadata with host-friendly field names."""

        value = self.client.read("Converter")
        if not isinstance(value, dict):
            raise PowerTestBenchError("metadata readback is not a ThingSet object")

        result = {}
        for public_name, (thingset_name, _) in self._METADATA_FIELDS.items():
            if thingset_name not in value:
                raise PowerTestBenchError(
                    f"Converter/{thingset_name} is missing from the readback"
                )
            result[public_name] = value[thingset_name]
        return result

    @staticmethod
    def _validate_metadata_value(name, value, maximum_length):
        if not isinstance(value, str):
            raise ValueError(f"{name} must be a string")
        if not value or len(value) > maximum_length:
            raise ValueError(
                f"{name} must contain 1 to {maximum_length} ASCII characters"
            )
        if any(ord(character) < 0x20 or ord(character) > 0x7E for character in value):
            raise ValueError(f"{name} must contain printable ASCII only")
        return value

    def set_metadata(
        self,
        *,
        board_name=None,
        board_version=None,
        serial_number=None,
        firmware_version=None,
    ):
        """Set one or more persistent converter identity fields."""

        supplied = {
            "board_name": board_name,
            "board_version": board_version,
            "serial_number": serial_number,
            "firmware_version": firmware_version,
        }
        values = {}
        for public_name, value in supplied.items():
            if value is None:
                continue
            thingset_name, maximum_length = self._METADATA_FIELDS[public_name]
            values[thingset_name] = self._validate_metadata_value(
                public_name, value, maximum_length
            )
        if not values:
            raise ValueError("at least one metadata field must be supplied")

        readback = self._write_and_verify("Converter", values)
        return {
            public_name: readback[thingset_name]
            for public_name, (thingset_name, _) in self._METADATA_FIELDS.items()
        }

    def power_on(
        self,
        leg,
        *,
        connect_driver=False,
        connect_capacitor=False,
        **leg_settings,
    ):
        """Safely configure and energize exactly one selected leg."""

        leg_number = self._normalize_leg(leg)
        self._validate_bool("connect_driver", connect_driver)
        self._validate_bool("connect_capacitor", connect_capacitor)
        forbidden = {"enable", "driver", "capacitor"} & set(leg_settings)
        if forbidden:
            names = ", ".join(sorted(forbidden))
            raise ValueError(f"power_on controls these settings directly: {names}")

        try:
            self.set_mode(TesterMode.POWER_OFF)
            self.configure_leg(1, enable=False)
            self.configure_leg(2, enable=False)
            self.configure_leg(
                leg_number,
                enable=True,
                driver=connect_driver,
                capacitor=connect_capacitor,
                **leg_settings,
            )
            self.set_mode(TesterMode.POWER_ON)
        except Exception as exc:
            try:
                self.shutdown()
            except Exception:
                pass
            if isinstance(exc, PowerTestBenchError):
                raise
            raise PowerTestBenchError(f"power-on sequence failed: {exc}") from exc

    def shutdown(self, *, disconnect_hardware=False):
        """Request POWER_OFF and disable both legs, attempting every step."""

        self._validate_bool("disconnect_hardware", disconnect_hardware)
        failures = []

        try:
            self.set_mode(TesterMode.POWER_OFF)
        except Exception as exc:
            failures.append(str(exc))

        for leg_number in (1, 2):
            settings = {"enable": False}
            if disconnect_hardware:
                settings.update(driver=False, capacitor=False)
            try:
                self.configure_leg(leg_number, **settings)
            except Exception as exc:
                failures.append(str(exc))

        if failures:
            raise PowerTestBenchError(
                "shutdown completed with errors: " + "; ".join(failures)
            )
